src/pip/_internal/wheel.py

```diff
@@ -773,52 +773,84 @@ def _contains_egg_info(
+def should_build(
     req,  # type: InstallRequirement
+    need_wheel,  # type: bool
     format_control,  # type: FormatControl
-    should_unpack,  # type: bool
-    cache_available  # type: bool
 ):
-    # type: (...) -> Optional[bool]
     """
-    Return whether to build an InstallRequirement object using the
-    ephemeral cache.
-
-    :param cache_available: whether a cache directory is available for the
-        should_unpack=True case.
-
-    :return: True or False to build the requirement with ephem_cache=True
-        or False, respectively; or None not to build the requirement.
+    Return whether a wheel should be built based on the requirement
+    characteristics, and if a wheel is ultimately needed or not.
     """
     if req.constraint:
         # never build requirements that are merely constraints
-        return None
+        return False
+
     if req.is_wheel:
-        if not should_unpack:
+        if need_wheel:
             logger.info(
                 'Skipping %s, due to already being wheel.', req.name,
             )
-        return None
-    if not should_unpack:
-        # i.e. pip wheel, not pip install;
-        # return False, knowing that the caller will never cache
-        # in this case anyway, so this return merely means "build it".
-        # TODO improve this behavior
+        # never build a requirement that is already a wheel
         return False
 
-    if req.editable or not req.source_dir:
-        return None
+    if need_wheel:
+        return True
+
+    if req.editable:
+        # don't build an editable if no wheel is needed
+        return False
+
+    if not req.source_dir:
+        # TODO explain what no source_dir means
+        return False
 
     if "binary" not in format_control.get_allowed_formats(
             canonicalize_name(req.name)):
         logger.info(
             "Skipping bdist_wheel for %s, due to binaries "
             "being disabled for it.", req.name,
         )
-        return None
+        return False
+
+    return True
+
+
+def should_use_ephemeral_cache(
+    req,  # type: InstallRequirement
+    format_control,  # type: FormatControl
+    cache_available,  # type: bool
+):
+    # type: (...) -> Optional[bool]
+    """
+    Return whether to build an InstallRequirement object using the
+    ephemeral cache.
+
+    :param cache_available: whether a cache directory is available for the
+        should_unpack=True case.
+
+    :return: True or False to build the requirement with ephem_cache=True
+        or False, respectively; or None not to build the requirement.
+    """
+    if req.editable:
+        # don't cache editable requirements because they can
+        # be locally modified
+        return True
+
+    if not req.source_dir:
+        # TODO explain what no source_dir means
+        return True
+
+    if "binary" not in format_control.get_allowed_formats(
+            canonicalize_name(req.name)):
+        # TODO explain this
+        return True
```

Comment:
This I'm not convinced. I see benefit to have should_build and should_cache independent: it's easier to reason about them. For instance if you look at the comment for `if req.editable`, there is a different reason to not build and to not cache editables. If we reintroduce a `should_build` test in `should_cache`, I think we pay a big price in readability to remove some maybe redundant tests that may be only accidentally the same.

Of course the fact that I don't fully understand the "no source_dir" and "no binary" parts does not help :) Yet I suspect the reasoning for caching and building in those cases could be different too.

Let me know what you think after reading the latest version.

Reply:
@cjerdonek I gave one more thought to this and I now think this test `need wheel and not should_build` it's actually not equivalent. In case `need_wheel` is False it would skip the test and may end up, e.g. caching an editable in install mode. Tss what a rabbit hole I put myself in :)

So for now I'm quite happy with the current version.

Reply:
>  I gave one more thought to this and I now think this test need wheel and not should_build it's actually not equivalent. In case need_wheel is False it would skip the test and may end up, e.g. caching an editable in install mode.

I don't see any issue.. What I suggested was logically the same by design. In install mode with an editable, `should_build()` would have returned `False`, so `should_use_ephemeral_cache()` would never have been called. To reiterate my comment, I don't think we should copy-paste all those lines verbatim. If it turns out that the logic needs to be different at some point in the future, we can always reconsider (but until then, YAGNI, as they say). If you'd prefer not to call `should_build()` with different arguments inside `should_cache()`, another possibility would be to refactor those copy-pasted lines into a helper function that is called from both functions. You can also document that `should_cache()` should only be called when `should_build()` is known to be true.